src/net/sourceforge/kolmafia/textui/Parser.java

```diff
@@ -4461,6 +4475,44 @@ public static Location mergeLocations(final Command start, final Command end) {
 
   // **************** Parse errors *****************
 
+  public class AshDiagnostic {
+    final Location location;
+    final DiagnosticSeverity severity;
+    final String message;
+    final List<String> additionalMessages;
+
+    private AshDiagnostic(
+        final Location location,
+        final DiagnosticSeverity severity,
+        final String message,
+        final String... additionalMessages) {
+      this.location = location;
+      this.severity = severity;
+      this.message = message;
+      this.additionalMessages = new ArrayList<>();
+      for (final String additionalMessage : additionalMessages) {
+        this.additionalMessages.add(additionalMessage);
+      }
+    }
+
+    public String toString() {
+      StringBuilder result = new StringBuilder();
+
+      result.append(this.message);
+      result.append(" (");
+
+      result.append(Parser.getFileAndRange(Parser.this.shortFileName, this.location.getRange()));
+
+      result.append(")");
+
+      for (final String additionalMessage : this.additionalMessages) {
+        result.append(" " + additionalMessage);
```

Comment:
This ends up copying additionalMessage twice: once to construct " " + additionalMessage, and a second time to append it to result.

Since you're already using a StringBuilder, instead consider:

    result.append(" ");
    result.append(additionalMessage);